DMA Serial Reading to Circular buffer

rondlh opened the pull request:
VERY EXPERIMENTAL BUILD THAT USES DMA CIRCULAR READING TO THE RX SERIAL BUFFER, NO ISR NEEDED.

The Dma Controller Just Does Its Work In The Background And Keeps Copying The Received Serial Characters To The Rx Buffer.
Marlin Will Read And Process The Received Data, Freeing Up The Buffer.
Emergency Parser Is Working Too, But A Bit Different Than Before.

By Default Marlin Uses Interrupt Based Serial Reading Provided By The Arduino Platform.
On My Setup This Causes Data Corruption (Lost Bytes) On My Setup Starting At 250k Baud When Stressing The System.
In Many Cases Data Corruption Can Go Unnoticed, Because The Resulting Gcode Command Is Still Valid.
I Only Found This Data Corruption Issue When I Started Using Crcs For The Gcodes.
I Believe This Is An Issue In The Arduino Platform.

With Dma Serial Reading I Can Use 1m Baud Without Any Issues.
Dma Is Of Course Hardware Dependent. I Only Implemented Stm32f2/F4. 
Test Hardware Used: Mks Monster 8 V1.0 With Btt Tft35v3.0 (Ahead Of Current Build).
resolves: #26322

rondlh committed:
Update README.md

EXPERIMENTAL BUILD

rondlh committed:
DMA SERIAL READING TO CIRCULAR RX BUFFER

STM32F2/F4 only

rondlh committed:
Revert "Update README.md"

This reverts commit fbbf539321a727b5617138d10a0265daf148dc69.

rondlh commented:
This is a tech demo only... for STM32F2 and F4.
The default serial reading that Marlin uses is based on Arduino's interrupt based reading.
I found that this causes data corruption, some bytes are dropped once in a while, but usually that goes unnoticed because the resulting gcode commands are usually still valid. When I started using CRCs for the gcode the problem became obvious, I have 1 or 2 lost bytes per hour at 250K baud. Much less at 115200 baud, but errors still occur (only under stress, high command rates).
I think there is an issue/bug in the Arduino serial reading.
With this DMA based serial reading I can use 1M baud, and found it to be rock solid. Even a specific stress test I use with over 1400 commands per second works fine when using DMA reading.

Used hardware: MKS Monster 8 V1.0 with BTT TFT3.5 V3.0
Only minor changes are required for Marlin, replace the (Arduino) HardwareSerial class with HardwareSerial2, which is located in 2 additional files.
![DMA_WRITE](https://github.com/MarlinFirmware/Marlin/assets/77279634/5bb9bcd5-d0d4-45a0-89f3-5b283b8d57e5)

rondlh committed:
Changed HAVE_HWSERIALx to USING_HW_SERIALx

Changed HAVE_HWSERIAL1 which is an internal Arduino variable to USING_HW_SERIAL1 so the serial port gets selected automatically.

ETE-Design commented:
@rondlh Will this be possible to add to STM32F1 Also? Having some MKS Robin Lite Boards, so would love to add this feature to them 😀

rondlh commented:
> @rondlh Will this be possible to add to STM32F1 Also? Having some MKS Robin Lite Boards + BTT TFT3.5 V3.0, so would love to add this feature to them 😀

Yes, no problem at all, but I don't have the hardware, it just takes a few renames of the register actually (like mentioned above), and perhaps change "channel" to "stream" or something... If you are willing to test the code then I will give it a try. OK?

thinkyhead committed:
formatting

thinkyhead committed:
HardwareSerial2 => HAL_HardwareSerial

thinkyhead committed:
misc.

thinkyhead commented on Marlin/src/HAL/STM32/HardwareSerial.h:
```diff
@@ -0,0 +1,67 @@
+  USART_TypeDef * uart;
+  uint32_t dma_rcc;
+  uint32_t dma_channel;
+  DMA_Stream_TypeDef * dma_streamRX;
```
Some `#include` is needed above this struct for the header that defines `DMA_Stream_TypeDef`.

rondlh replied:
Yes, fixed

rondlh committed:
Added STM32F1xx support, untested!

Untested, it compiles, but not tested. I don't have STM32F1xx hardware.

rondlh committed:
Merge branch 'bugfix-2.1.x' of https://github.com/rondlh/Marlin-DMA_Reading into bugfix-2.1.x

rondlh commented:
I made an update for STM32F1 support... experimental! It compiles, but I don't have the hardware.
@ETE-Design Needs testing!
[STM32-HardwareSerial2.zip](https://github.com/MarlinFirmware/Marlin/files/12841810/STM32-HardwareSerial2.zip)
[STM32-HardwareSerial2.zip](https://github.com/MarlinFirmware/Marlin/files/12841811/STM32-HardwareSerial2.zip)

ETE-Design commented:
> I made an update for STM32F1 support... experimental! It compiles, but I don't have the hardware. @ETE-Design Needs testing! [STM32-HardwareSerial2.zip](https://github.com/MarlinFirmware/Marlin/files/12841810/STM32-HardwareSerial2.zip)
> 
> [STM32F1-HardwareSerial2.zip](https://github.com/MarlinFirmware/Marlin/files/12841818/STM32F1-HardwareSerial2.zip)

Is there anything I need to do on the BTT TFT3.5 V3.0 other that load newest firmware from their Github?

rondlh committed:
Merging STM32F1 update with Scott's good work, convert to HAL_HardwareSerial

I hope I didn't erase too many of your corrections

rondlh commented:
> @rondlh Is there anything I need to do on the BTT TFT3.5 V3.0 other that load newest firmware from their Github?

@ETE-Design 
No FW update required, this is completely independent of the serial software you use to send gcode to your motherboard.
Please test if the communication between the STM32F1 motherboard and TFT still works, if so then things should be fine.
It's difficult to make code "blindly"... so good luck. Please let me know :D

Please delete the code in your post above... it's already outdated. Lot's of support ongoing...

Here is the latest STM32F1 code, don't forget to delete the old **HardwareSerial2** files!!!
[src-HAL-STM32F1 MarlinSerial.zip](https://github.com/MarlinFirmware/Marlin/files/12841912/src-HAL-STM32F1.MarlinSerial.zip)
[src-HAL-STM32-MarlinSerial.zip](https://github.com/MarlinFirmware/Marlin/files/12841916/src-HAL-STM32-MarlinSerial.zip)

I also use a BTT TFT35 V3.0, and have been struggling to find the cause of the data corruption. This Marlin update solves the issue. Stable at 1M Baud, before 250K was very unstable. The TFT CRC feature I use on the TFT is not released yet, no PR even. I can send it to you for testing if you want. If you think having a CRC option on the TFT is useful, then please go to the BTT Touch Screen github, look into the issues and leave a nice message for me... I opened a feature request for this, but there is not much support for it up till now. I think it is very useful, because it will show data corruption immediately, without it a lot of data corruption can go unnoticed.

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26328

thinkyhead commented:
Please, at the very least fix the formatting by hand and make sure it compiles before pushing changes. Otherwise it's left to me to fix the formatting by hand so that I can begin to make sense of the changes.

thinkyhead committed:
format and fix

thinkyhead committed:
etc

thinkyhead commented:
The HAL / STM32F1 is based on Maple which is quite different from HAL / STM32 which is based on the code supplied by STMicro. It is unlikely that code for one HAL will work for the other without a large number of changes to interoperate with the Maple libraries. We consider Maple to be deprecated although we keep it around for testing, so there is no need to do further development on HAL / STM32F1 except to fix bugs.

rondlh commented:
@thinkyhead It compiles both on STM32F1(untested) and F2/F4 (working). I had to undo one change to make it compile (move ""protected").
The formatting and changes you made should be preserved, I did my best to keep everything you did, but I might have made a mistake left or right. I appreciate your support! I'm sorry if I causes any issues, you are just so fast... I will continue tomorrow.

EvilGremlin commented:
@rodh i think you dont get it, STM32 HAL is for all ST chips, STM32F1 HAL only used for separate _maple envs.

rondlh commented:
> @rodh i think you dont get it, STM32 HAL is for all ST chips, STM32F1 HAL only used for separate _maple envs.

You could be right... so you mean I should convert the code to HAL calls, so all STM chips are supported?
Could you point me in the right direction? I don't know what "maple envs" are.

thinkyhead committed:
cleanup. best guess for USART4

EvilGremlin commented:
NVM i'm blind, missed your switch in STM32 HAL. Still doesn't make much sense to add same changes to _maple (yet) as nowadayas it's mostly usef fior chinese not-clone chips which tend to have differences in DMA

thinkyhead committed:
etc

thinkyhead committed:
Leave STM32F1 (Maple) as-is

thinkyhead committed:
fix ctor

thinkyhead commented on Marlin/src/HAL/STM32/MarlinSerial.cpp:
```diff
@@ -87,11 +87,12 @@
   TERN_(EMERGENCY_PARSER, _serial.rx_callback = _rx_callback);
 }
 
+// IRON, IT SEEMS THIS CODE IS NEVER USED (STM32)??????????
```
There are a few references to `_rx_complete_irq` in the code. For example:
```cpp
  if (baud == WIFI_BAUDRATE)
    uart_attach_rx_callback(&_serial, _rx_complete_irq);
  else
    USART1->CR1 |= USART_CR1_RE;  // Preserve word length, etc. Use 'or' to preserve USART_CR1_M_8N1
```
```cpp
#define DECLARE_SERIAL_PORT(ser_num) \
  void _rx_complete_irq_ ## ser_num (serial_t * obj); \
  MSerialT MSerial ## ser_num (true, USART ## ser_num, &_rx_complete_irq_ ## ser_num); \
  void _rx_complete_irq_ ## ser_num (serial_t * obj) { MSerial ## ser_num ._rx_complete_irq(obj); }
```

rondlh replied:
> There are a few references to _rx_complete_irq in the code:

Yes, I saw, but I never enable the interrupt callbacks, so it never gets called when using DMA serial reading. This is what I want, so I don't need to change the code there.

thinkyhead commented:
> It compiles both on STM32F1(untested) and F2/F4 (working).

The changes are breaking builds besides your limited case. For example `mftest -tchitu_f103`. As long as this is still just experimental to play around with, that's fine. But it should be complete and ready to go, not breaking any other builds, before it can be widely reviewed or merged for the wider world.

rondlh committed:
Don't break other build, keep old code if not STM32F1 F2 F4

rondlh commented:
> The changes are breaking builds besides your limited case. For example `mftest -tchitu_f103`. As long as this is still just experimental to play around with, that's fine. But it should be complete and ready to go, not breaking any other builds, before it can be widely reviewed or merged for the wider world.

I noticed the code breaks 2 builds, let me check what's going on. I don't know how to check all builds locally, but I can try the 2 failing builds of course.

rondlh commented:
**Build test: chitu_F103**
env: mks_robin_lite3_maple   board: BOARD_MKS_ROBIN_LITE3
Fails because in Configuration.h "SERIAL_PORT 0"
HAL.h:97:3: error: static assertion failed: SERIAL_PORT must be from 1 to 5. You can also use -1 if the board supports Native USB.
It builds fine when I set "SERIAL_PORT 1".

`Build test: LERDGEK`
env: LERDGEK_usb_flash_drive   board: BOARD_LERDGE_K
Fails because of the same reason as chitu_F103
It builds fine when I set "SERIAL_PORT 1".

I'm not sure how to solve these build errors, anybody?

rondlh committed:
RX and TX buffers are required to make this work

I didn't touch TX, but still added a safety to make sure serial writing can always work

ETE-Design commented:
@thinkyhead Now that HAL / STM32F1 use Maple, is it anyhow possible to just let it use HAL / STM32 like the rest of the STM32 boards?
I'm using the MKS_Robin_Lite who use the STM32F1

ellensp commented:
@ETE-Design not related to this PR....

Using env:mks_robin_lite_maple uses Maple
Using env:mks_robin_lite uses uses stm32duino, ie HAL / STM32

ETE-Design commented:
@ellensp Your right on that, was only asking because @thinkyhead mentioned that STM32 on The STM32F1 is using Maple, and would like to get rid of that, but think I missed something then, cause it can also be running on HAL /  STM32 like other STM32 Boards right?

rondlh committed:
Support F1 DMA Structure in the same file

rondlh committed:
Set SERIAL_PORT 1 for testing

rondlh committed:
Corrected issues with STM32F1

Added flush

rondlh commented:
@ETE-Design I made some correction for the STM32F1 platform.
Please try:
default_envs = mks_robin_lite
Board: BOARD_MKS_ROBIN_LITE

[src-HAL-STM32-HardwareSerial.zip](https://github.com/MarlinFirmware/Marlin/files/12845082/src-HAL-STM32-HardwareSerial.zip)

These files go in the src\HAL\STM32 directory. Please restore your str\HAL\STM32F1 to it's original state (no HardwareSerial.cpp/h, and original MarlinSerial.cpp/h)

rondlh committed:
Corrected DMA start register for STM32F1

EvilGremlin commented:
I could've test F1, but my F1 board is on CDC now...

rondlh commented:
> I could've test F1, but my F1 board is on CDC now...

Reported working on MKS Robin Pro V1.0 (STM32F103), also fine on my MKS Monster 8 V1.0 (STM32F407)

rondlh committed:
SERIAL_PORT set back to 0

rondlh committed:
Disable Arduino to resolve variable naming conflicts

Removed #include "Arduino.h"

rondlh commented:
```
default_envs = chitu_f103
#define MOTHERBOARD BOARD_CHITU3D_V5
#define MOTHERBOARD BOARD_CHITU3D_V6
#define MOTHERBOARD BOARD_CHITU3D_V9
```
All compile fine for "SERIAL_PORT 1"
Fails for "SERIAL_PORT 0" because of
`Marlin\src\HAL\STM32\../../inc/../HAL/../HAL/STM32/HAL.h:67:4: error: #error "SERIAL_PORT  must be from 1 to 9, or -1 for Native USB."`

**How to resolve this issue?
Anybody knows why so many tests get cancelled?**

EvilGremlin commented:
All tests are cancelled after three fails i think?
chitu_f103 test pulls example conf from repo (as you can see) and that conf has correct serial port. No idea why it would fail, maybe CI not working properly again.

rondlh commented:
> All tests are cancelled after three fails i think? chitu_f103 test pulls example conf from repo (as you can see) and that conf has correct serial port. No idea why it would fail, maybe CI not working properly again.

Thanks for your feedback!
Any other platforms you think are useful to support?
I checked how many boards there are for different MCUs.
2 x GD32F1
2 x STM32F7
8 x STM32G0
8 x STM32H7

The improvement in data reading reliability is really amazing when using DMA. Online I found lot about of stories about difficult boards and EMC causing data communication errors. Before 250K baud was not stable at all for me (STM32F407), now 1M baud is stable as a rock. I think there is something wrong in the Arduino platform.

EvilGremlin commented:
F0 too
also GD32F3 (cortex-m4)
And then HC32, AT32, N32 and whatever else they'll ship tomorrow...

> something wrong in the Arduino platform

DUH!

EvilGremlin commented:
Did you test with two serial ports? With both CDC and UART.

rondlh commented:
> Did you test with two serial ports? With both CDC and UART.

Yes, my standard testing is based in CDC with UART, so I can monitor what's happening and make sure no data is lost.
I also tested with 2 UARTS, the only thing you have to consider is that every UART has an RX and TX buffer, no problem as long as you have the memory available. The best way to test is sending Gcode with CRCs.

It would be quite easy to kick out Arduino for serial writing too and use STM32 interrupt based writing (or even DMA writing).

rondlh committed:
Guard Serial DMA writing code with "SERIAL_DMA"

rondlh commented:
Everyone, you now need to `#define SERIAL_DMA` to enjoy serial DMA reading.

rondlh committed:
STM32F0 and STM32F7 support, added Configuration_adv.h

More serial ports supported

rondlh committed:
Cleanup items not needed for SERIAL_DMA

rondlh committed:
Increase undefined or 0 size RX Buffer from 64 to 128 bytes

All supported boards have enough memory, 128 bytes is a safe minimum

rondlh committed:
Improve code readability, STM32F0 support

rondlh committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into bugfix-2.1.x

rondlh committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x

rondlh committed:
Merge branch 'bugfix-2.1.x' of https://github.com/rondlh/Marlin-DMA_Reading into bugfix-2.1.x

rondlh committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into bugfix-2.1.x

rondlh committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into bugfix-2.1.x

kisslorand commented:
> UPDATE: I got a report back that this one works on F1 and F2 and F4 :D

I tested this on a BTT SKR2 motherboard with STM32F429VTG6 MCU and it fails. It is connected to a BTT TFT35 v3 TFT. As soon as the TFT sends anything the motherboard resets. The motherboard reports the reason of reset as "Watchdog Reset".

![other_0](https://github.com/MarlinFirmware/Marlin/assets/50251547/c1c12f86-82d4-4376-8102-3cd0422251a7)

[Config Files.zip](https://github.com/MarlinFirmware/Marlin/files/12925252/Config.Files.zip)

rondlh committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into bugfix-2.1.x

narno2202 commented:
I test with PR #26074 ( last `FT_MOTION` PR with `FT_MOTION` enabled), and it works fine. It's not a surprise as I have the same motherboard and a BTT TFT35 E3 V3.0 (wth DMA firmware too but it's not mandatory). DMA brings me back to the early nineties with my 1040STE :)

rondlh commented:
> I test with PR #26074 ( last `FT_MOTION` PR with `FT_MOTION` enabled), and it works fine. It's not a surprise as I have the same motherboard and a BTT TFT35 E3 V3.0 (wth DMA firmware too but it's not mandatory). DMA brings me back to the early nineties with my 1040STE :)

Thanks for the feedback. They key benefit is a highly increased serial data reliability. Without it I noticed that sometimes a serial byte is dropped, which usually goes unnoticed because the resulting gcode is usually still valid.

danbrianwhite commented on Marlin/src/HAL/STM32/HardwareSerial.cpp:
```diff
@@ -0,0 +1,445 @@
+
+  #ifdef UART5 // must be F2 / F4 /F7
+    else if (peripheral == UART5) {
+      RX_DMA = { UART5, RCC_AHB1Periph_DMA1, 4, DMA1_Stream0 };
```
This fails on F1 device.

Can it be updated with something like
`RX_DMA = { UART4, RCC_AHBPeriph_DMA2, DMA2, DMA2_Channel5 };`

or `DMA2_Channel4`? I'm not sure what the current usage of these is in Marlin

rondlh replied:
> This fails on F1 device.

Thanks for the feedback. Which motherboard are you using?
What is the exact error message you get?

> Can it be updated with something like `RX_DMA = { UART4, RCC_AHBPeriph_DMA2, DMA2, DMA2_Channel5 };`

The UART4 code is present a few lines above already. A F1 device should not have a UART5, so the code should not cause issues. I could exclude the code for F1 devices if needed to solve this issue.

> or `DMA2_Channel4`? I'm not sure what the current usage of these is in Marlin

The DMA stream is linked to the UART port used, so that must be correct, it cannot freely be changed.

danbrianwhite replied:
I'm using `BOARD_MKS_ROBIN_E3_V1_1` and the issue is `DMA1_Stream0` does not exist for f1.

```
error: 'DMA1_Stream0' was not declared in this scope
```

narno2202 commented:
@rondlh , as you said, in the STM32F1 datasheet, only 4 UART are accessible by the DMA controller. In my opinion, others UART must be disabled,  which is not currently the case.

EvilGremlin commented:
UART numbers are quite nuanced and should be enabled according to serial ports in conf (and LCD_SERIAL_PORT). 
I.E. F401 UARTS are 1, 2 and 6.

rondlh commented:
@narno2202
> @rondlh , as you said, in the STM32F1 datasheet, only 4 UART are accessible by the DMA controller. In my opinion, others UART must be disabled, which is not currently the case.

Thanks, I agree, I will disable support for UART5 and USART6 for F0 and F1 because there is no DMA hardware support. I will push an update later today.

rondlh committed:
Fixed compilation error because UART5 has no DMA support on F0 and F1

Changed to "ANY" macro

rondlh committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into bugfix-2.1.x

thisiskeithb commented:
It'd be nice to add this for every MCU Marlin supports, but I know that's a tall order.

Getting STM32G0/G1* and STM32H7* supported in this PR would go a long way since those cover BTT's SKR Mini E3 V3.0, Manta, and Octopus series.

jmz52 commented:
For the record, so someone from the future can find this and use for troubleshooting.

`SERIAL_DMA` and `TFT_COLOR_UI`  might conflict in some hardware configuration - some UART and SPI devices use same DMA.
Example: DMA1_Channel5 on STM32F1 is used by USART1 and SPI2.

thinkyhead committed:
check for defines

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26328

thinkyhead commented:
The various DMA values are provided by the platform as defines, so I went ahead and adjusted the DMA / Stream conditions to check for those defines instead of specific STM32 models.

thinkyhead committed:
apply standards

thinkyhead committed:
update comments

thinkyhead committed:
add a test

thinkyhead committed:
Move some serial defaults

thinkyhead commented:
> Fails for "SERIAL_PORT 0" because of

We've long assumed that STM32 platforms number their serial ports starting from `MSerial1` to `MSerial9`. If that is no longer the case (or was never the case) then we should have a look at that.

thisiskeithb commented:
@rondlh: What's required for STM32G0/G1 and STM32H7 support? I have the hardware here (and we have the #testing channel on Discord) to test if you don't have the hardware.
